File: frontend/js/views/saas_magic_number.js
// SaaS magic number (sales efficiency), via /calc/saas-magic-number.
import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
const num = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 4 }));
export async function renderSaasMagicNumber(mount, _s) {
    const tok = currentViewToken(); if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.smn.h1.title">// SAAS MAGIC NUMBER</span></h1>
        <p class="muted small" data-i18n="view.smn.hint.intro">Sales efficiency: the quarter-over-quarter revenue increase, annualized (×4), divided by the prior quarter's sales & marketing spend. It measures how much new annual recurring revenue each dollar of S&M bought. Above ~0.75 is efficient enough to invest harder; below ~0.5 the go-to-market is not paying back. Not advice.</p>
        <div class="lpv-split"><div class="chart-panel"><h2 data-i18n="view.smn.h2.inputs">Quarterly revenue & spend</h2>
        <form id="smn-form" class="inline-form">
            <label><span data-i18n="view.smn.label.cur">Current quarter revenue ($)</span><input type="number" step="10000" min="0" name="current_quarter_revenue_usd" value="1100000" required></label>
            <label><span data-i18n="view.smn.label.prior">Prior quarter revenue ($)</span><input type="number" step="10000" min="0" name="prior_quarter_revenue_usd" value="1000000" required></label>
            <label><span data-i18n="view.smn.label.sm">Prior quarter S&M spend ($)</span><input type="number" step="10000" min="0" name="prior_quarter_sm_spend_usd" value="400000" required></label>
        </form></div><div id="smn-result" class="chart-panel lpv-preview"></div></div>`;
    applyUiI18n(mount);
    const form = mount.querySelector('#smn-form'); const n = (k) => Number(form.querySelector(`[name="${k}"]`).value) || 0;
    const gen = async () => {
        const body = { current_quarter_revenue_usd: n('current_quarter_revenue_usd'), prior_quarter_revenue_usd: n('prior_quarter_revenue_usd'), prior_quarter_sm_spend_usd: n('prior_quarter_sm_spend_usd') };
        try { const d = await api.calcSaasMagicNumber(body); if (!viewIsCurrent(tok)) return; res(mount, d); }
        catch (e) { showToast(e.message || t('view.smn.toast.error'), { level: 'error' }); }
    };
    const live = debounce(gen, 250); form.addEventListener('input', () => live()); form.addEventListener('submit', (e) => { e.preventDefault(); gen(); }); gen();
}
function res(mount, d) {
    const el = mount.querySelector('#smn-result');
    if (!d.valid) { el.innerHTML = `<p class="muted" data-i18n="view.smn.invalid">Prior quarter S&M spend must be positive.</p>`; applyUiI18n(el); return; }
    const cls = d.efficiency === 'efficient' ? 'pos' : (d.efficiency === 'poor' ? 'neg' : '');
    const effLabel = t('view.smn.eff.' + d.efficiency) || d.efficiency;
    el.innerHTML = `<div class="lpv-bar"><div class="cards">
        <div class="card ${cls}"><div class="label" data-i18n="view.smn.card.magic">Magic number</div><div class="value">${num(d.magic_number)}</div></div>
        <div class="card ${cls}"><div class="label" data-i18n="view.smn.card.eff">Efficiency</div><div class="value">${effLabel}</div></div>
        <div class="card"><div class="label" data-i18n="view.smn.card.arr">Annualized net-new ARR</div><div class="value">${money(d.annualized_net_new_arr_usd)}</div></div>
        <div class="card"><div class="label" data-i18n="view.smn.card.payback">S&M payback (months)</div><div class="value">${num(d.sm_payback_months)}</div></div>
    </div></div>`;
    applyUiI18n(el);
}

File: crates/traderview-core/src/saas_magic_number.rs
//! SaaS magic number — sales efficiency: how much new annual recurring revenue
//! each dollar of the prior period's sales & marketing spend generated. It is the
//! quarter-over-quarter increase in revenue, annualized (×4), divided by the prior
//! quarter's S&M spend. Above ~0.75 is efficient enough to invest harder; below
//! ~0.5 signals the go-to-market is not paying back. It also reports the implied
//! months to recover S&M. Distinct from the LTV/CAC and break-even ROAS modules.
//! Pure compute, not financial advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct MagicNumberInput {
    pub current_quarter_revenue_usd: f64,
    pub prior_quarter_revenue_usd: f64,
    /// Sales & marketing spend in the PRIOR quarter (it drives this quarter's growth).
    pub prior_quarter_sm_spend_usd: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq, Default)]
pub struct MagicNumberReport {
    /// (current − prior) × 4 — annualized net-new revenue.
    pub annualized_net_new_arr_usd: f64,
    /// Net-new ARR ÷ prior S&M spend.
    pub magic_number: f64,
    /// Efficiency band: "poor", "acceptable", or "efficient".
    pub efficiency: String,
    /// Implied months to recover the S&M spend (12 ÷ magic number).
    pub sm_payback_months: f64,
    pub valid: bool,
}

fn round4(x: f64) -> f64 {
    (x * 10_000.0).round() / 10_000.0
}

fn round2(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

pub fn generate(i: &MagicNumberInput) -> MagicNumberReport {
    if i.prior_quarter_sm_spend_usd <= 0.0 {
        return MagicNumberReport::default();
    }
    let net_new = (i.current_quarter_revenue_usd - i.prior_quarter_revenue_usd) * 4.0;
    let magic = net_new / i.prior_quarter_sm_spend_usd;
    let efficiency = if magic >= 0.75 {
        "efficient"
    } else if magic >= 0.5 {
        "acceptable"
    } else {
        "poor"
    };
    let payback = if magic > 0.0 { round2(12.0 / magic) } else { 0.0 };
    MagicNumberReport {
        annualized_net_new_arr_usd: round2(net_new),
        magic_number: round4(magic),
        efficiency: efficiency.to_string(),
        sm_payback_months: payback,
        valid: true,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> MagicNumberInput {
        MagicNumberInput {
            current_quarter_revenue_usd: 1_100_000.0,
            prior_quarter_revenue_usd: 1_000_000.0,
            prior_quarter_sm_spend_usd: 400_000.0,
        }
    }

    #[test]
    fn magic_one_efficient() {
        let d = generate(&base());
        assert!(close(d.annualized_net_new_arr_usd, 400_000.0));
        assert!(close(d.magic_number, 1.0));
        assert_eq!(d.efficiency, "efficient");
        assert!(close(d.sm_payback_months, 12.0));
    }

    #[test]
    fn high_growth_high_magic() {
        let d = generate(&MagicNumberInput { current_quarter_revenue_usd: 1_300_000.0, ..base() });
        assert!(close(d.magic_number, 3.0));
        assert_eq!(d.efficiency, "efficient");
    }

    #[test]
    fn poor_efficiency_band() {
        // Small growth, large spend → magic < 0.5.
        let d = generate(&MagicNumberInput { current_quarter_revenue_usd: 1_030_000.0, ..base() });
        // (30k×4)/400k = 0.3.
        assert!(close(d.magic_number, 0.3));
        assert_eq!(d.efficiency, "poor");
    }

    #[test]
    fn acceptable_band() {
        let d = generate(&MagicNumberInput { current_quarter_revenue_usd: 1_060_000.0, ..base() });
        // (60k×4)/400k = 0.6.
        assert!(close(d.magic_number, 0.6));
        assert_eq!(d.efficiency, "acceptable");
    }

    #[test]
    fn zero_spend_invalid() {
        let d = generate(&MagicNumberInput { prior_quarter_sm_spend_usd: 0.0, ..base() });
        assert!(!d.valid);
    }
}
